### PR TITLE
pam_systemd_home: close sd-bus connection on failure

### LINKED ISSUE #31375: Another occurrence of `(sd-pam)[475]: PAM Attempted to close sd-bus after fork whose connection is opened before the fork, this should not happen.`
Spotted in one of the recent PRs in TEST-46-HOMED:

```
[  998.281831] systemd[1]: Started mysshserver@1-127.0.0.1:4711-127.0.0.1:56114.service.
[  998.281870] systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/mysshserver_401_2d127_2e0_2e0_2e1_3a4711_2d127_2e0_2e0>
[  998.281894] systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/mysshserver_401_2d127_2e0_2e0_2e1_3a4711_2d127_2e0_2e0>
[  998.281917] systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1 interface=org.freedesktop.systemd1.Manager member=JobRemove>
[  998.281941] systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/mysshserver_401_2d127_2e0_2e0_2e1_3a4711_2d127_2e0_2e0>
[  998.281964] systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/mysshserver_401_2d127_2e0_2e0_2e1_3a4711_2d127_2e0_2e0>
[  998.281988] systemd-executor[490]: SELinux enabled state cached to: disabled
[  998.323945] (sshd)[490]: Found cgroup2 on /sys/fs/cgroup/, full unified hierarchy
[  998.324015] (sd-pam)[475]: pam_systemd_home(systemd-user:session): Failed to release user home: Access denied
[  998.324239] systemd-homed[48]: Got notification that all sessions of user homedsshtest ended, deactivating automatically.
[  998.324347] (sd-pam)[475]: pam_unix(systemd-user:session): session closed for user homedsshtest
[  998.324380] systemd-homed[48]: Successfully closed pin fd on home for homedsshtest.
[  998.324410] (sd-pam)[475]: pam_systemd(systemd-user:session): Failed to release session: Access denied
[  998.324446] systemd-homed[48]: Sending to worker: {"luksDiscard":true,"luksPbkdfType":"pbkdf2","luksPbkdfTimeCostUSec":1000,"enforcePasswordPolicy":false,"stopDelayUS>
[  998.324546] (sd-pam)[475]: pam_warn(systemd-user:setcred): function=[pam_sm_setcred] flags=0x4 service=[systemd-user] terminal=[] user=[homedsshtest] ruser=[<unknown>>
[  998.324586] systemd-homed[48]: Successfully forked off '(sd-homework)' as PID 491.
[  998.324613] (sd-pam)[475]: PAM Attempted to close sd-bus after fork whose connection is opened before the fork, this should not happen.
[  998.324640] systemd-homed[48]: homedsshtest: changing state active → deactivating
```

Full journal: https://mrc0mmand.fedorapeople.org/journals/close-sd-bus-after-fork.journal

Caught by a check introduced in https://github.com/systemd/systemd/pull/31032.

### PR DESCRIPTION
**Pull request by yuwata**
Hopefully fixes #31375.

cc @mrc0mmand

### CODE REVIEW

**Comment by mrc0mmand** [comment]
Hmpf, I can still reproduce the issue with this patch (but now it at least has some possibly useful information):

```
[163569.473163] H systemd[717]: Reached target exit.target.
[163569.473213] H systemd[717]: Exit.
[163569.473428] H systemd[1]: mysshserver@1-127.0.0.1:4711-127.0.0.1:41062.service: Forked /usr/sbin/sshd as 732 (via CLONE_INTO_CGROUP)
[163569.473500] H systemd[1]: Closing set fd 38 (socket:[379453])
[163569.473542] H systemd[1]: Closing set fd 35 (socket:[393310])
[163569.473616] H systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/mysshserver_401_2d127_2e0_2e0_2e1_3a4711_2d127_2e0_2e0_2e1_3a41062_2eservice interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=825>
[163569.473696] H systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/mysshserver_401_2d127_2e0_2e0_2e1_3a4711_2d127_2e0_2e0_2e1_3a41062_2eservice interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=826>
[163569.473743] H systemd[1]: mysshserver@1-127.0.0.1:4711-127.0.0.1:41062.service: Changed dead -> running
[163569.473798] H systemd[1]: mysshserver@1-127.0.0.1:4711-127.0.0.1:41062.service: Job 506 mysshserver@1-127.0.0.1:4711-127.0.0.1:41062.service/start finished, result=done
[163569.473857] H systemd[1]: Started mysshserver@1-127.0.0.1:4711-127.0.0.1:41062.service.
[163569.474635] H systemd[717]: Bus bus-system: changing state RUNNING → CLOSED
[163569.476607] H systemd[1]: Got message type=method_call sender=:1.1 destination=org.freedesktop.systemd1 path=/org/freedesktop/systemd1/job/504 interface=org.freedesktop.DBus.Properties member=Get  cookie=56 reply_cookie=0 signature=ss error-name=n/a error-message=n/a
[163569.476679] H systemd[1]: Sent message type=method_return sender=n/a destination=:1.1 path=n/a interface=n/a member=n/a cookie=847 reply_cookie=56 signature=v error-name=n/a error-message=n/a
[163569.477464] H (sshd)[732]: Found cgroup2 on /sys/fs/cgroup/, full unified hierarchy
[163569.507489] H systemd-homed[48]: Got notification that all sessions of user homedsshtest ended, deactivating automatically.
[163569.507640] H (sd-pam)[718]: pam_systemd_home(systemd-user:session): New sd-bus connection (system-bus-pam-systemd-home-718) opened.
[163569.509149] H systemd-homed[48]: Successfully closed pin fd on home for homedsshtest.
[163569.509394] H (sd-pam)[718]: pam_systemd_home(systemd-user:session): Failed to release user home: Access denied
[163569.509576] H systemd-homed[48]: Sending to worker: {"luksDiscard":true,"luksPbkdfType":"pbkdf2","luksPbkdfTimeCostUSec":1000,"enforcePasswordPolicy":false,"stopDelayUSec":0,"userName":"homedsshtest","perMachine":[{"diskSize":0,"rebalanceWeight":0,"matchMachineId":"6b>
[163569.509760] H (sd-pam)[718]: pam_unix(systemd-user:session): session closed for user homedsshtest
[163569.509856] H systemd-homed[48]: Successfully forked off '(sd-homework)' as PID 733.
[163569.509908] H (sd-pam)[718]: pam_systemd(systemd-user:session): New sd-bus connection (system-bus-pam-systemd-718) opened.
[163569.509954] H systemd-homed[48]: homedsshtest: changing state active → deactivating
[163569.510004] H (sd-pam)[718]: pam_systemd(systemd-user:session): Failed to release session: Access denied
[163569.510133] H systemd-homed[48]: Got message type=method_call sender=:1.86 destination=org.freedesktop.home1 path=/org/freedesktop/home1 interface=org.freedesktop.home1.Manager member=ReleaseHome  cookie=2 reply_cookie=0 signature=s error-name=n/a error-message=n/a
[163569.510217] H (sd-pam)[718]: pam_warn(systemd-user:setcred): function=[pam_sm_setcred] flags=0x4 service=[systemd-user] terminal=[] user=[homedsshtest] ruser=[<unknown>] rhost=[<unknown>]
[163569.510273] H systemd-homed[48]: Sent message type=method_call sender=n/a destination=org.freedesktop.DBus path=/org/freedesktop/DBus interface=org.freedesktop.DBus member=GetConnectionUnixUser cookie=249 reply_cookie=0 signature=s error-name=n/a error-message=n/a
[163569.510355] H (sd-pam)[718]: PAM Attempted to close sd-bus (system-bus-pam-systemd-home-717) after fork whose connection is opened before the fork, this should not happen.
```

Full journal: https://mrc0mmand.fedorapeople.org/journals/close-sd-bus-after-fork-with-31403.journal

**Comment by yuwata** [comment]
I guess I found the root cause. @mrc0mmand Please test again.

File: src/home/pam_systemd_home.c [hunk 1]
```diff
@@ -826,6 +826,12 @@ _public_ PAM_EXTERN int pam_sm_open_session(
         if (r != PAM_SUCCESS)
                 return r;
 
+        /* Explicitly get saved PamBusData here. Otherwise, this function may succeed without setting 'd'
+         * even if there is an opened sd-bus connection, and it will be leaked. See issue #31375. */
+        r = pam_get_bus_data(handle, "pam-systemd-home", &d);
```
Comment (YHNdnzj) [inline src/home/pam_systemd_home.c:831]: Hmm, with this, it's no longer necessary for `acquire_home` or `acquire_user_record` to return `PamBusData`?

Comment (yuwata) [reply src/home/pam_systemd_home.c:831]: It's necessary, as at this time there may be no bus connection, and `d` is NULL in that case. And a new bus connection may be opened in acquire_home() later below.

Comment (YHNdnzj) [reply src/home/pam_systemd_home.c:831]: Indeed... But it's easier to follow if there's only one source of truth. So maybe just introduce a cleanup func that does `pam_get_bus_data` + `pam_bus_data_disconnectp`?

**Review by mrc0mmand — APPROVED** [review]
The issue indeed seems to be gone now, thank you!

File: src/home/pam_systemd_home.c (hunk 1 above)
Comment (yuwata) [reply src/home/pam_systemd_home.c:831]: Hm, but on failure, we (and hopefully other user applications) will call `pam_setcred(PAM_DELETE_CRED)` soon, and the bus connection will be closed there. So, the explicit close in this function is not necessary.
Such hack (`pam_get_bus_data` + `pam_bus_data_disconnectp`) is only necessary when this function succeeds without calling `pam_acquire_bus_connection()`. I think that is overkill at least now. If we will introduce more PAM modules in the future with the similar pitfalls, then let's reconsider the suggestion.

**Comment by yuwata** [comment]
Upgrading the green label. See comments in the above.

File: src/home/pam_systemd_home.c (hunk 1 above)
Comment (YHNdnzj) [reply src/home/pam_systemd_home.c:831]: > Such hack (pam_get_bus_data + pam_bus_data_disconnectp) is only necessary when this function succeeds without calling pam_acquire_bus_connection(). 

I generally agree, but `pam_get_bus_data` is now unconditionally called, so it kinda feels weird to me.

But doesn't matter too much after all.

**Pull request merged**